src/uct/cuda/gdr_copy/gdr_copy_md.c

```diff
@@ -259,36 +300,246 @@ static uct_md_ops_t md_ops = {
+    uct_gdr_copy_rcache_region_t *region;
+
+    UCT_MD_MEM_DEREG_CHECK_PARAMS(params, 0);
+
+    region = ucs_container_of(params->memh, uct_gdr_copy_rcache_region_t, memh);
+    ucs_rcache_region_put(md->rcache, &region->super);
+    return UCS_OK;
+}
+
+static uct_md_ops_t md_rcache_ops = {
+    .close               = uct_gdr_copy_md_close,
+    .query               = uct_gdr_copy_md_query,
+    .mkey_pack           = uct_gdr_copy_mkey_pack,
+    .mem_reg             = uct_gdr_copy_mem_rcache_reg,
+    .mem_dereg           = uct_gdr_copy_mem_rcache_dereg,
+    .detect_memory_type  = ucs_empty_function_return_unsupported,
+};
+
+static ucs_status_t
+uct_gdr_copy_rcache_mem_reg_cb(void *context, ucs_rcache_t *rcache,
+                               void *arg, ucs_rcache_region_t *rregion,
+                               uint16_t rcache_mem_reg_flags)
+{
+    uct_gdr_copy_md_t *md = context;
+    int *flags            = arg;
+    uct_gdr_copy_rcache_region_t *region;
+
+    if (rcache_mem_reg_flags & UCS_RCACHE_MEM_REG_HIDE_ERRORS) {
+        *flags |= UCT_MD_MEM_FLAG_HIDE_ERRORS;
+    }
+
+    region = ucs_derived_of(rregion, uct_gdr_copy_rcache_region_t);
+    return uct_gdr_copy_mem_reg_internal(&md->super, (void*)region->super.super.start,
+                                         region->super.super.end -
+                                         region->super.super.start,
+                                         *flags, &region->memh);
+}
+
+static void uct_gdr_copy_rcache_mem_dereg_cb(void *context, ucs_rcache_t *rcache,
+                                             ucs_rcache_region_t *rregion)
+{
+    uct_gdr_copy_md_t *md = context;
+    uct_gdr_copy_rcache_region_t *region;
+
+    region = ucs_derived_of(rregion, uct_gdr_copy_rcache_region_t);
+    (void)uct_gdr_copy_mem_dereg_internal(&md->super, &region->memh);
+}
+
+static void uct_gdr_copy_rcache_dump_region_cb(void *context, ucs_rcache_t *rcache,
+                                               ucs_rcache_region_t *rregion, char *buf,
+                                               size_t max)
+{
+    uct_gdr_copy_rcache_region_t *region = ucs_derived_of(rregion,
+                                                          uct_gdr_copy_rcache_region_t);
+    uct_gdr_copy_mem_t *memh = &region->memh;
+
+    snprintf(buf, max, "bar ptr:%p", memh->bar_ptr);
+}
+
+static ucs_rcache_ops_t uct_gdr_copy_rcache_ops = {
+    .mem_reg     = uct_gdr_copy_rcache_mem_reg_cb,
+    .mem_dereg   = uct_gdr_copy_rcache_mem_dereg_cb,
+    .merge       = (void*)ucs_empty_function,
+    .dump_region = uct_gdr_copy_rcache_dump_region_cb
+};
+
+static ucs_status_t
+uct_gdr_copy_md_init(uct_gdr_copy_md_t *md, uct_component_t *component,
+                     const uct_gdr_copy_md_config_t *md_config, int locked)
+{
+    ucs_rcache_params_t rcache_params;
+    ucs_status_t status;
+
     md->super.component = &uct_gdr_copy_component;
     md->reg_cost        = md_config->uc_reg_cost;
 
+    if (locked) {
```

Comment:
maybe pass `ops` as a parameter, instead of the `locked` param - will be more explicit

Reply:
can't we deduce "locked" from md_config?